### PR TITLE
Spark 3.3: Add years transform function

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds `years` transform function to Spark 3.3.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @kbendick @rdblue @RussellSpitzer @szehon-ho @flyrain @gaborkaszab @nastra

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.years(date)";
+    }
+
+    @Override
+    public Integer produceResult(InternalRow input) {
+      // return null for null input to match what Spark does in codegen
```
Comment (RussellSpitzer) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java:101]: Makes sense, as long as this is what Spark does 👍

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  @Test
+  public void testThatMagicFunctionsAreInvoked() {
+    String dateValue = "date('2017-12-01')";
+    Assertions.assertThat(scalarSql("EXPLAIN EXTENDED SELECT system.years(%s)", dateValue))
+        .asString()
+        .isNotNull()
+        .contains(
+            "staticinvoke(class org.apache.iceberg.spark.functions.YearsFunction$DateToYearsFunction");
```
Comment (RussellSpitzer) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java:107]: nit: should we get the package and function name programmatically? Probably not a big deal as I can't imagine this changing

**Review by RussellSpitzer — APPROVED** [review]
Looks good to me, added a suggestion for some testing and there is one typo that needs to be cleaned up in a test.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java (hunk 1 above)
Comment (aokolnychyi) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java:101]: Yep, that's what we decided to do in other functions such as `bucket` and `truncate`.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    Assert.assertEquals(
+        "Expected to produce 2017 - 1970 = 47",
+        47,
+        scalarSql("SELECT system.years(date('2017-12-01'))"));
+    Assert.assertEquals(
+        "Expected to produce 1970 - 1970 = 0",
+        0,
+        scalarSql("SELECT system.years(date('1970-01-01'))"));
+    Assert.assertEquals(
+        "Expected to produce 19690 - 1970 = -1",
```
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java:47]: Oops, will fix.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java (hunk 2 above)
Comment (aokolnychyi) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java:107]: Sounds more reliable, I'll try.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+public class TestSparkYearsFunction extends SparkTestBaseWithCatalog {
+
+  @Before
+  public void useCatalog() {
+    sql("USE %s", catalogName);
+  }
+
+  @Test
+  public void testDates() {
+    Assert.assertEquals(
```
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java:38]: I was inspired by our existing tests in core but let me add one more test for this too.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      return new DataType[] {DataTypes.TimestampType};
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.years(timestamp)";
+    }
+
+    @Override
+    public Integer produceResult(InternalRow input) {
```
Comment (nastra) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java:124]: nit: this method appears to be the same for TimestampToYearsFunction + DateToYearsFunction, should we maybe move it to BaseToYearsFunction?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+    StructField valueField = inputType.fields()[0];
+    DataType valueType = valueField.dataType();
+
+    if (valueType instanceof DateType) {
+      return new DateToYearsFunction();
+    } else if (valueType instanceof TimestampType) {
+      return new TimestampToYearsFunction();
+    } else {
+      throw new UnsupportedOperationException("Expected value to be date or timestamp");
```
Comment (nastra) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java:54]: nit: should we print what the underlying type of `inputType` was?

**Review by nastra — APPROVED** [review]
overall LGTM, just some tiny nits/suggestions

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java]: Fixed.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java (hunk 2 above)
Comment (aokolnychyi) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java]: Changed.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java (hunk 5 above)
Comment (aokolnychyi) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java:113]: They are slightly different. One fetches integers and the other one fetches longs from `input`.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java (hunk 6 above)
Comment (aokolnychyi) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java]: Good idea, added.

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @RussellSpitzer @nastra!

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java (hunk 5 above)
Comment (nastra) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/YearsFunction.java:113]: ah good call, I missed this tiny detail :)

**Pull request merged**